src/smtencoding/FStar.SMTEncoding.Encode.fst

```diff
@@ -789,30 +789,52 @@ let encode_top_level_let :
                          [dummy_var], app
                     else vars, maybe_curry_fvb (FStar.Syntax.Util.range_of_lbname lbn) fvb (List.map mkFreeV vars)
                 in
-                let pat, app, (body, decls2) =
-                  let is_logical =
-                    match (SS.compress t_body).n with
-                    | Tm_fvar fv when S.fv_eq_lid fv FStar.Parser.Const.logical_lid -> true
-                    | _ -> false
+                let is_logical =
+                  match (SS.compress t_body).n with
+                  | Tm_fvar fv when S.fv_eq_lid fv FStar.Parser.Const.logical_lid -> true
+                  | _ -> false
+                in
+                let is_smt_theory_symbol =
+                    let fv = FStar.Compiler.Util.right lbn in
+                    Env.fv_has_attr env.tcenv fv FStar.Parser.Const.smt_theory_symbol_attr_lid
+                in
+                let should_encode_logical =
+                    not is_smt_theory_symbol
+                    && (quals |> List.contains Logic || is_logical)
+                in
+                let make_eqn name pat app body =
+                    //NS 05.25: This used to be mkImp(mk_and_l guards, mkEq(app, body))),
+                    //But the guard is unnecessary given the pattern
+                    Util.mkAssume(mkForall (U.range_of_lbname lbn)
+                                           ([[pat]], vars, mkEq(app,body)),
+                                  Some (BU.format1 "Equation for %s" (string_of_lid flid)),
+                                  (name ^ "_" ^ fvb.smt_id))
+                in
+                let eqns,decls2 =
+                  let basic_eqn_name =
+                    if should_encode_logical
+                    then "defn_equation"
+                    else "equation"
```

Comment:
We'll always encode a "basic equation", that's the non-logical encoding

But, preserving names so as not to break hints etc. When we're also doing a logical encoding, the basic equation is called a "defn_equation", i.e., the equation that provides the raw definition.